_posts/2024-04-10-benchmarks-over-time.md

```diff
@@ -0,0 +1,548 @@
+The higher the variety of analyses in the benchmark, the more broadly useful the tool can be.
+
+#### What scale of data can it handle?
+
+Many benchmarks are deliberately smaller than typical workloads. 
+This allows the benchmark to complete in a reasonable amount of time when run with many configurations. 
+However, an important question to answer when selecting a system is whether the size of your data can be handled within the size of your compute resources.
+
+### Limitations of Benchmarking over Time
+
+There are some limitations when looking at the performance of a system over time.
+If a feature is brand new, there is no prior performance to compare to!
+As a result, this post focuses on fundamental workloads rather than DuckDB's ever-increasing set of integrations with different lakehouse data formats, cloud services, and more.
+
+The code used to run the benchmark also avoids many of DuckDB's [Friendlier SQL](https://duckdb.org/docs/guides/sql_features/friendly_sql) additions, as those have also been added more recently. 
+(When writing these queries, it felt like going back in time!)
+
+
+## Benchmark Design
+
+### Summary
+
+* H2O.ai, plus import/export and window function tests
+* Python instead of R
+* 5GB scale for everything, plus 50GB scale for group bys and joins
+* Median of 3 runs
+* Using a MacBook Pro M1 with 16GB RAM
+* DuckDB Versions 0.2.7 through 1.0.0
+    * Nearly 3 years, from 2021-06-14 to 2024-06-03
+* Default settings
+
+
+### H2O.ai as the Foundation
+
+This post measures DuckDB's performance over time on the H2O.ai benchmark for both joins and group by queries.
+Please see our previous [blog](https://duckdb.org/2023/04/14/h2oai.html) [posts](https://duckdb.org/2023/11/03/db-benchmark-update.html) for details on why we believe the H2O.ai benchmark is a good approach!
+
+The result of each H2O.ai query is written to a table in a persistent DuckDB file.
+This does require additional work when compared with an in-memory workflow (especially the burden on the SSD rather than RAM), but improves scalability and is a common approach for larger analyses.
+
+As in the current H2O.ai benchmark, categorical-type columns (`VARCHAR` columns with low cardinality) were converted to the `ENUM` type as a part of the benchmark.
+The time for converting into `ENUM` columns was included in the benchmark time, and resulted in a lower total amount of time (so the upfront conversion was worthwhile).
+However, the `ENUM` data type was not fully operational in DuckDB until version 0.6.1, so earlier versions skip this step.
+
+### Python Client
+
+To measure interoperability with other dataframe formats, we have used Python rather than R (used by H2O.ai) for this analysis.
+We do continue to use R for the data generation step for consistency with the benchmark.
+Python is DuckDB's most popular client, so this is also the most representative of real world performance.
```

Comment:
```suggestion
Python is DuckDB's most popular client, and it's my preferred tool to hack things together.```
Or equivalent, I would not put down an objective reason (that looks a bit weak to me) but go with something that says that writing Python is easier on you.

Reply:
Fair point! I can make that change - it was a significant factor... :-)

Reply:
Adjusted! Thanks!